### PR TITLE
[QOL] Candy Upgrade Notification Rework

### PR DESCRIPTION
**Pull request by josericardo-fo**
## What are the changes?

Reworked the Candy Upgrade Notification logic to take account of the new "Buy Egg" option.

## Why am I doing these changes?

A new mechanic was implemented (Buying Eggs).

## What did change?

The logic of the 'On' setting 

### Before
image

### After
image

## How to test the changes?

Test if the candy icon appears whenever you can buy eggs.

## Checklist
- [X] There is no overlap with another PR?
- [X] The PR is self-contained and cannot be split into smaller PRs?
- [X] Have I provided a clear explanation of the changes?
- [X] Have I tested the changes (manually)?
    - [X] Are all unit tests still passing? (`npm run test`)
- [X] Are the changes visual?
  - [X] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/locales/zh_TW/settings.ts [hunk 1]
```diff
@@ -43,7 +43,7 @@ export const settings: SimpleTranslationEntries = {
   "moveAnimations": "招式動畫",
   "showStatsOnLevelUp": "升級時顯示能力值",
   "candyUpgradeNotification": "糖果升級提示",
-  "passivesOnly": "僅被動",
+  "passivesFirst": "Passives First", // Changed from "Passives Only" to "Passives First", needs new translation
```
Comment (40chyan) [inline src/locales/zh_TW/settings.ts:46]:
```suggestion
  "passivesFirst": "優先被動", // Changed from "Passives Only" to "Passives First", needs new translation
```

File: src/locales/zh_CN/settings.ts [hunk 2]
```diff
@@ -43,7 +43,7 @@ export const settings: SimpleTranslationEntries = {
   "moveAnimations": "招式动画",
   "showStatsOnLevelUp": "升级时显示能力值",
   "candyUpgradeNotification": "糖果升级提示",
-  "passivesOnly": "仅被动",
+  "passivesFirst": "Passives First", // Changed from "Passives Only" to "Passives First", needs new translation
```
Comment (40chyan) [inline src/locales/zh_CN/settings.ts:46]:
```suggestion
  "passivesFirst": "优先被动", // Changed from "Passives Only" to "Passives First", needs new translation
```

Comment (josericardo-fo) [reply src/locales/zh_CN/settings.ts:46]: make sure to remove the comment

File: src/locales/fr/settings.ts [hunk 3]
```diff
@@ -43,7 +43,7 @@ export const settings: SimpleTranslationEntries = {
   "moveAnimations": "Animations de combat",
   "showStatsOnLevelUp": "Afficher stats à la montée de N.",
   "candyUpgradeNotification": "Notif amélioration de bonbon",
```
Comment (Adri1) [inline src/locales/fr/settings.ts:45]:
```suggestion
  "candyUpgradeNotification": "Notif. amélio. de bonbon",
```

File: src/locales/fr/settings.ts [hunk 4]
```diff
@@ -43,7 +43,7 @@ export const settings: SimpleTranslationEntries = {
...
-  "passivesOnly": "Passifs",
+  "passivesFirst": "Passives First", // Changed from "Passives Only" to "Passives First", needs new translation
```
Comment (Adri1) [inline src/locales/fr/settings.ts:46]:
```suggestion
  "passivesFirst": "Prio Passif",
```

**Review by Adri1 — COMMENTED** [review]
Added French + adjustment of an exiting entry to make it fit)

File: src/locales/de/settings.ts [hunk 5]
```diff
@@ -43,7 +43,7 @@ export const settings: SimpleTranslationEntries = {
   "moveAnimations": "Attacken Animationen",
   "showStatsOnLevelUp": "Werte beim Aufleveln anzeigen",
   "candyUpgradeNotification": "Bonbon Upgrade Benachrichtigung",
-  "passivesOnly": "Nur Passive",
+  "passivesFirst": "Passives First", // Changed from "Passives Only" to "Passives First", needs new translation
```
Comment (CodeTappert) [inline src/locales/de/settings.ts:46]:
```suggestion
  "passivesFirst": "Passive zuerst",
```

File: src/locales/ko/settings.ts [hunk 6]
```diff
@@ -43,7 +43,7 @@ export const settings: SimpleTranslationEntries = {
   "moveAnimations": "배틀 애니메이션",
   "showStatsOnLevelUp": "레벨업 능력치 표시",
   "candyUpgradeNotification": "사탕 업그레이드 알림",
-  "passivesOnly": "패시브만",
+  "passivesFirst": "Passives First", // Changed from "Passives Only" to "Passives First", needs new translation
```
Comment (EnochG1) [inline src/locales/ko/settings.ts:46]:
```suggestion
  "passivesFirst": "패시브우선",
```

**Review by KimJeongSun — COMMENTED** [review]
Overall, LGTM
As a minor suggestion, the menu name `passive first` seems to omit the actual logic that valueReduction is also prioritized.
`prioritize Passive and value reduction` appears to be a clearer description, but it doesn't seem concise enough for a menu name, so it's hard to recommend.

**Comment by josericardo-fo** [comment]
Thanks!
If someone suggests a better name, I'll gladly change it. But I can't really find another concise name for that setting.

**Review by KimJeongSun — APPROVED** [review]

**Review by EnochG1 — APPROVED** [review]

**Comment by NicusPulcis** [comment]
@josericardo-fo i've made an italian translation

File: src/ui/starter-select-ui-handler.ts [hunk 7]
```diff
@@ -893,6 +905,18 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+  /**
+   * Determines if a passive upgrade is maxed for the given species ID
+   * @param speciesId The ID of the species to check the passive of
+   * @returns true if the passive has been maxed
+   */
+  isValueReductionMaxed(speciesId: number): boolean {
+    // Get this species ID's starter data
+    const starterData = this.scene.gameData.starterData[speciesId];
+
+    return starterData.valueReduction >= 2;
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:917]: Magic numbers 👎 
```suggestion
    return starterData.valueReduction >= VALUE_REDUCTION_MAX;
```
(variable has to be defined somewhere central. See if there is other code occurrences for the exact number

File: src/ui/starter-select-ui-handler.ts [hunk 8]
```diff
@@ -957,15 +981,19 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
         }
       ],};
 
+    const passiveUnlocked = this.isPassiveUnlocked(species.speciesId);
     const passiveAvailable = this.isPassiveAvailable(species.speciesId);
-    // 'Only Passives' mode
+    const valueReductionMaxed = this.isValueReductionMaxed(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:986]:
```suggestion
    const isValueReductionMaxed = this.isValueReductionMaxed(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

File: src/ui/starter-select-ui-handler.ts [hunk 9]
```diff
@@ -957,15 +981,19 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    const valueReductionAvailable = this.isValueReductionAvailable(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:987]:
```suggestion
    const isValueReductionAvailable = this.isValueReductionAvailable(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

File: src/ui/starter-select-ui-handler.ts [hunk 10]
```diff
@@ -957,15 +981,19 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    const eggAvailable = this.isSameSpeciesEggAvailable(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:988]:
```suggestion
    const isEggAvailable = this.isSameSpeciesEggAvailable(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

File: src/ui/starter-select-ui-handler.ts [hunk 11]
```diff
@@ -989,16 +1017,20 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
       return;
     }
 
+    const passiveUnlocked = this.isPassiveUnlocked(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:1020]:
```suggestion
    const isPassiveUnlocked = this.isPassiveUnlocked(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

File: src/ui/starter-select-ui-handler.ts [hunk 12]
```diff
@@ -989,16 +1017,20 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
     const passiveAvailable = this.isPassiveAvailable(species.speciesId);
-    // 'Only Passive Unlocks' mode
+    const valueReductionMaxed = this.isValueReductionMaxed(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:1022]:
```suggestion
    const isValueReductionMaxed = this.isValueReductionMaxed(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

File: src/ui/starter-select-ui-handler.ts [hunk 13]
```diff
@@ -989,16 +1017,20 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    const valueReductionAvailable = this.isValueReductionAvailable(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:1023]:
```suggestion
    const isValueReductionAvailable = this.isValueReductionAvailable(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

File: src/ui/starter-select-ui-handler.ts [hunk 14]
```diff
@@ -989,16 +1017,20 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
...
+    const eggAvailable = this.isSameSpeciesEggAvailable(species.speciesId);
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:1024]:
```suggestion
    const isEggAvailable = this.isSameSpeciesEggAvailable(species.speciesId);
```
`is` (or `has`) prefix helps to quickly distinguish a boolean

**Review by flx-sta — CHANGES_REQUESTED** [review]

File: src/ui/starter-select-ui-handler.ts (hunk 7 above)
Comment (josericardo-fo) [reply src/ui/starter-select-ui-handler.ts:917]: it is not defined anywhere 💀

Comment (flx-sta) [reply src/ui/starter-select-ui-handler.ts:917]: I know. Thats part of your task

File: src/ui/starter-select-ui-handler.ts [hunk 15]
```diff
@@ -268,6 +268,7 @@ export default class StarterSelectUiHandler extends MessageUiHandler {
   private classicWinIcons: Phaser.GameObjects.Image[];
   private candyUpgradeIcon: Phaser.GameObjects.Image[];
   private candyUpgradeOverlayIcon: Phaser.GameObjects.Image[];
+  private valueReductionMax: 2;
```
Comment (flx-sta) [inline src/ui/starter-select-ui-handler.ts:271]: It shouldn't be an attribute of StartSelectUiHandler. It should be a constant defined outside of the class. for other code files to be available if needed (exported)

**Review by flx-sta — CHANGES_REQUESTED** [review]

File: src/ui/starter-select-ui-handler.ts (hunk 15 above)
Comment (josericardo-fo) [reply src/ui/starter-select-ui-handler.ts:271]: I don't really know where to define that. The costs of everything are located in this file as a const.

image

Comment (josericardo-fo) [reply src/ui/starter-select-ui-handler.ts:271]: maybe like this? lol

image

Comment (flx-sta) [reply src/ui/starter-select-ui-handler.ts:271]: yeah like that but since it's an actual constant name it `export const VALUE_REDUCTION_MAX = 2`

Comment (josericardo-fo) [reply src/ui/starter-select-ui-handler.ts]: fixed

**Review by Adri1 — APPROVED** [review]

**Review by CodeTappert — APPROVED** [review]

**Review by flx-sta — APPROVED** [review]
lgtm 🚀  (code review only)

**Comment by josericardo-fo** [comment]
@brain-frog @bennybroseph Take a look whenever you can, we implemented this, so you guys maybe want to give an opinion.

**Comment by 40chyan** [comment]
@Tempo-anon this PR all good

**Comment by josericardo-fo** [comment]
Moved changes to #3279

**Pull request closed without merging**